src/FINDOLOGIC/Api/RequestBuilders/RequestBuilder.php

```diff
@@ -0,0 +1,297 @@
+<?php
+
+namespace FINDOLOGIC\Api\RequestBuilders;
+
+use FINDOLOGIC\Api\Definitions\Endpoint;
+use FINDOLOGIC\Api\Definitions\QueryParameter;
+use FINDOLOGIC\Api\Exceptions\InvalidParamException;
+use FINDOLOGIC\Api\Exceptions\ParamNotSetException;
+use FINDOLOGIC\Api\Client;
+use FINDOLOGIC\Api\Config;
+use FINDOLOGIC\Api\Validators\ParameterValidator;
+use InvalidArgumentException;
+use Valitron\Validator;
+
+abstract class RequestBuilder
+{
+    const
+        SET_VALUE = 'set',
+        ADD_VALUE = 'add';
+
+    /** @var array */
+    protected $params = [];
+
+    /** @var array */
+    protected $requiredParams = [
+        QueryParameter::SHOP_URL,
+    ];
+
+    /** @var string */
+    protected $endpoint;
+
+    /** @var Config */
+    protected $config;
+
+    /** @var Client */
+    protected $client;
+
+    /**
+     * Sends a request to FINDOLOGIC. This may include an alivetest if the requested resource is a search or
+     * a navigation request.
+     */
+    abstract public function sendRequest();
+
+    public function __construct(Config $config)
+    {
+        $this->config = $config;
+        $this->client = new Client($config);
+    }
+
+    /**
+     * Builds the request URL based on the set params.
+     *
+     * @return string
+     */
+    public function buildRequestUrl()
+    {
+        $shopUrl = $this->getParam(QueryParameter::SHOP_URL);
+        // If the shopkey was not manually overridden, we take the shopkey from the config.
+        if (!isset($this->getParams()[QueryParameter::SHOPKEY])) {
+            $this->params['shopkey'] = $this->config->getShopkey();
+        }
+        $queryParams = http_build_query($this->params);
+        // Removes indexes from query params. E.g. attrib[0] will be attrib[].
```

Comment:
Why? It shouldn't pose a problem. If it does, explain what the problem is.

Reply:
I think I had an issue with properties regarding that. Attributes should be fine, this comment should just illustrate that all parameter indexes will be dropped.

Reply:
When something like this is necessary, readers of the code should be well-informed regarding the motivation to do so. "I think" is insufficient - please verify that a problem exists and describe it in a fashion that someone reading this code for the first time can understand. If it turns out not to be a problem, we can remove it.